scio-core/src/main/scala/com/spotify/scio/package.scala

```diff
@@ -78,6 +78,12 @@ package object scio {
       Await.result(self.flatMap(identity), atMost)
   }
 
+  /** Set the name of the next transform. */
+  implicit class NamedTfSCollection[T: ClassTag](val wrapped: SCollection[T]) {
+    def withName(tfName: String): SCollection[T] =
```

Comment:
Probably true, you might not even need a `NamedSCollectionImpl`, just reflect on `self.nameProvider` should be sufficient.